File: SelfHomeApp/app/src/main/java/it/dani/selfhomeapp/middle/persistence/Persistence.java
package it.dani.selfhomeapp.middle.persistence;

import android.util.Log;

import androidx.annotation.NonNull;

import org.jetbrains.annotations.NotNull;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketTimeoutException;
import java.net.UnknownHostException;
import java.util.Arrays;
import java.util.Set;
import java.util.TreeSet;

import it.dani.selfhomeapp.middle.ipersistence.IDevicePersistence;
import it.dani.selfhomeapp.middle.ipersistence.IGroupPersistence;

public class Persistence implements IDevicePersistence, IGroupPersistence {
    private static InetAddress ADDRESS;
    private static int PORT;

    public static Persistence getPersitence(@NotNull String addr, int port) throws UnknownHostException {
        if(port < 0 || port > 65535) throw  new IllegalArgumentException("Error: invalid port value");
        ADDRESS = InetAddress.getByName(addr);
        PORT = port;

        return new Persistence();
    }

    @NonNull
    @Override
    public String freeText(@NonNull String text)
    {
        String result;

        try
        {
            result = this.communicate(text,PersistenceTimeout.INFINITE.valueOf());

        } catch (IOException e){
            Log.e("getDeviceFromGroup","Error: " + e.getMessage());
            result = "";
        }

        return result;
    }

    @Override
    public boolean toggleDevice(@NonNull String deviceName)
    {
        boolean result = false;

        try
        {
            String resultS = this.communicate("SET DISP <" + deviceName + "> !(GET DISP <" + deviceName + ">);");
            if(resultS != null)
            {
                result = Boolean.parseBoolean(resultS);
            }

        } catch (IOException | NumberFormatException e){
            Log.e("getDeviceFromGroup","Error: " + e.getMessage());
            result = false;
        }

        return result;
    }

    @Override
    public Set<String> getDevices() {
        Set<String> result = new TreeSet<>();

        try {
            String response = this.communicate("LST");

            if(response.length() > 0)
            {
                result.addAll(Arrays.asList(response.split(";")));
            }

        } catch (SocketTimeoutException e) {
            throw new SelfHomePersistenceException(e);

        } catch (IOException e){
            Log.e("getDevice","Error: " + e.getMessage());
        }

        return result;
    }

    @Override
    public Set<String> getGroups() {
        Set<String> result = new TreeSet<>();

        try
        {
            String response = this.communicate("LGR");

            if(response.length() > 0)
            {
                result.addAll(Arrays.asList(response.split(";")));
            }

        } catch (SocketTimeoutException e) {
            throw new SelfHomePersistenceException(e);

        } catch (IOException e){
            Log.e("getGroups","Error: " + e.getMessage());
        }

        return result;
    }

    @Override
    public Set<String> getDeviceFromGroup(@NonNull String groupName) {
        Set<String> result = new TreeSet<>();

        try
        {
            String response = this.communicate("LDG <" + groupName + ">");

            if(response.length() > 0)
            {
                result.addAll(Arrays.asList(response.split(";")));
            }

        } catch (IOException e){
            Log.e("getDeviceFromGroup","Error: " + e.getMessage());
        }

        return result;
    }

    @Override
    public boolean addDeviceToGroup(@NonNull String deviceName, @NonNull String groupName) {
        boolean flagErr = true;

        try
        {
            //flagErr = !this.communicate("ADG <" + deviceName + "> <" + groupName + ">").equals("OK");
            flagErr = !Boolean.parseBoolean(this.communicate("ADG <" + deviceName + "> <" + groupName + ">;"));
        } catch (IOException e) {
            Log.e("addDeviceToGroup","Error: " + e.getMessage());
        }

        return !flagErr;
    }

    @Override
    public boolean delDeviceToGroup(@NonNull String deviceName, @NonNull String groupName) {
        boolean flagErr = true;

        try
        {
            //flagErr = !this.communicate("DDG <" + deviceName + "> <" + groupName + ">").equals("OK");
            flagErr = !Boolean.parseBoolean(this.communicate("DDG <" + deviceName + "> <" + groupName + ">;"));
        } catch (IOException e) {
            Log.e("delDeviceToGroup","Error: " + e.getMessage());
        }

        return !flagErr;
    }

    @Override
    public boolean setStateDevice(@NonNull String deviceName, int state) {
        boolean flagErr = true;

        try
        {
            //flagErr = !this.communicate("SET DISP <" + deviceName + "> " + state).equals("OK");
            flagErr = !Boolean.parseBoolean(this.communicate("SET DISP <" + deviceName + "> " + state + ";"));
        } catch (IOException e) {
            Log.e("setDeviceState","Error: " + e.getMessage());
        }

        return !flagErr;
    }

    @Override
    public boolean setStateGroup(@NonNull String groupName, int state) {
        boolean flagErr = true;

        try
        {
            //flagErr = !this.communicate("SET GRUP <" + groupName + "> <" + state + ">").equals("OK");
            flagErr = !Boolean.parseBoolean(this.communicate("SET GRUP <" + groupName + "> <" + state + ">;"));
        } catch (IOException e) {
            Log.e("setStateGroup","Error: " + e.getMessage());
        }

        return !flagErr;
    }

    @Override
    public boolean addNewGroup(@NonNull String groupName) {
        boolean flagErr = true;

        try
        {
            //flagErr = !this.communicate("NGR <" + groupName + ">").equals("OK");
            flagErr = !Boolean.parseBoolean(this.communicate("NGR <" + groupName + ">;"));
        } catch (IOException e) {
            Log.e("addNewGroup","Error: " + e.getMessage());
        }

        return !flagErr;
    }

    @Override
    public boolean delGroup(@NonNull String groupName)
    {
        boolean flagErr = true;

        try
        {
            //flagErr = !this.communicate("DGR <" + groupName + ">").equals("OK");
            flagErr = !Boolean.parseBoolean(this.communicate("DGR <" + groupName + ">;"));
        } catch (IOException e) {
            Log.e("delGroup","Error: " + e.getMessage());
        }

        return !flagErr;
    }

    @Override
    public int getDeviceState(@NonNull String deviceName) {
        int result;

        try
        {
            String resultS = this.communicate("GET DISP <" + deviceName + ">;");
            if(resultS.equals("OFF")) result = 0;
            else result = Integer.parseInt(resultS);

        } catch (IOException | NumberFormatException e) {
            Log.e("getDeviceState","Error: " + e.getMessage());

            throw new SelfHomePersistenceException(e);
        }

        return result;
    }

    private String communicate(@NonNull String str) throws IOException
    {
        return this.communicate(str,PersistenceTimeout.STANDARD.valueOf());
    }

    private String communicate(@NonNull String str, int timeout) throws IOException
    {
        StringBuilder result = new StringBuilder();

        try(DatagramSocket sock = new DatagramSocket())
        {
            byte[] buffer = new byte[1024];
            DatagramPacket pack = new DatagramPacket(buffer,buffer.length,ADDRESS,PORT);
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            OutputStreamWriter osw = new OutputStreamWriter(baos);
            osw.write(str);
            osw.flush();
            pack.setData(baos.toByteArray());

            sock.send(pack);
            sock.setSoTimeout(timeout);

            pack.setData(buffer);
            sock.receive(pack);
            InputStreamReader isr = new InputStreamReader(new ByteArrayInputStream(pack.getData()));

            char car;
            while((car = (char)isr.read()) > 0)
            {
                result.append(car);
            }

        }

        return result.toString();
    }
}

File: SelfHomeApp/app/src/main/java/it/dani/selfhomeapp/middle/persistence/Persistence.kt
package it.dani.selfhomeapp.middle.persistence

import android.util.Log
import it.dani.selfhomeapp.middle.ipersistence.IDevicePersistence
import it.dani.selfhomeapp.middle.ipersistence.IGroupPersistence
import java.io.*
import java.lang.StringBuilder
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.util.*
import kotlin.IllegalArgumentException

class Persistence private constructor(private val address: InetAddress, private val port: Int) : IDevicePersistence, IGroupPersistence {

    init {
        if(port <= 0 || port > 65535) throw IllegalArgumentException("Port must be positive")
    }

    override fun freeText(text: String): String {
        return try {
            this.communicate(text)

        } catch (e : IOException) {
            Log.e("freeText","Error: ${e.message}")
            ""
        }
    }

    override fun toggleDevice(deviceName: String): Boolean {
        return try {
            val response = this.communicate("TGL DISP <${deviceName}>;")
            if(response.isNotEmpty()) {
                response.toBoolean()
            } else {
                false
            }
        } catch (e : IOException) {
            Log.e("toggleDevice","Error: ${e.message}")
            false
        }
    }

    override fun getDevices(): MutableSet<String> {
        val result = TreeSet<String>()

        try {
            val response = this.communicate("LST;")
            if(response.isNotEmpty()) {
                result += response.split(";")
            }

        } catch (e : IOException) {
            Log.e("getDevice","Error: ${e.message}")
        }

        return result
    }

    override fun setStateDevice(deviceName: String, state: Int): Boolean {
        return try {
            this.communicate("SET DISP <${deviceName}> ${state};").lowercase(Locale.getDefault()).toBooleanStrict()
        } catch (e : IOException) {
            Log.e("setDeviceState","Error: ${e.message}")
            false
        }
    }

    override fun getDeviceState(deviceName: String): Int {
        try {
            val result = this.communicate("GET DISP <${deviceName}>;")
            return if (result == "OFF") {
                0
            } else {
                result.toInt()
            }
        } catch(e : IOException) {
            Log.e("getDeviceState","Error: ${e.message}")
            throw SelfHomePersistenceException()   //TODO inserisci e
        }
    }

    override fun getGroups(): MutableSet<String> {
        val result = TreeSet<String>()

        try {
            val response = this.communicate("LGR;")
            if(response.isNotEmpty()) {
                result += response.split(";")
            }

        } catch (e : IOException) {
            Log.e("getGroups","Error: ${e.message}")
        }

        return result
    }

    override fun setStateGroup(groupName: String, state: Int): Boolean {
        return try {
            this.communicate("SET GRUP <${groupName}> <${state}>;").lowercase(Locale.getDefault()).toBooleanStrict()
        } catch (e : IOException) {
            Log.e("setStateGroup","Error: ${e.message}")
            false
        }
    }

    override fun addNewGroup(groupName: String): Boolean {
        return try {
            this.communicate("NGR <${groupName}>;").lowercase(Locale.getDefault()).toBooleanStrict()
        } catch (e : IOException) {
            Log.e("addNewGroup","Error: ${e.message}")
            false
        }
    }

    override fun delGroup(groupName: String): Boolean {
        return try {
            this.communicate("DGR <${groupName}>;").lowercase(Locale.getDefault()).toBooleanStrict()
        } catch (e : IOException) {
            Log.e("delGroup","Error: ${e.message}")
            false
        }
    }

    override fun getDeviceFromGroup(groupName: String): MutableSet<String> {
        val result = TreeSet<String>()

        try {
            val response = this.communicate("LDG <${groupName}>;")
            if(response.isNotEmpty()) {
                result += response.split(";")
            }

        } catch (e : IOException) {
            Log.e("getDeviceFromGroup","Error: ${e.message}")
        }

        return result
    }

    override fun addDeviceToGroup(deviceName: String, groupName: String): Boolean {
        return try {
            this.communicate("ADG <${deviceName}> <${groupName}>;").lowercase(Locale.getDefault()).toBooleanStrict()
        } catch (e : IOException) {
            Log.e("addDeviceToGroup","Error: ${e.message}")
            false
        }
    }

    override fun delDeviceToGroup(deviceName: String, groupName: String): Boolean {
        return try {
            this.communicate("DDG <${deviceName}> <${groupName}>;").lowercase(Locale.getDefault()).toBooleanStrict()
        } catch (e : IOException) {
            Log.e("delDeviceToGroup","Error: ${e.message}")
            false
        }
    }

    private fun communicate(str : String) : String {
        return this.communicate(str,Timeout.STANDARD)
    }

    private fun communicate(str : String, timeout : Timeout) : String {
        val sock = DatagramSocket()

        val buffer = ByteArray(1024)

        val pack = DatagramPacket(buffer,buffer.size,this.address,this.port)
        val baos = ByteArrayOutputStream()
        val osw = OutputStreamWriter(baos)

        osw.write(str)
        osw.flush()
        pack.data = baos.toByteArray()

        sock.send(pack)
        sock.soTimeout = timeout.value

        pack.data = buffer
        sock.receive(pack)
        sock.close()

        val dis = DataInputStream(ByteArrayInputStream(pack.data))
        val result = StringBuilder()

        var car = dis.read()
        while (car > 0) {
            result.append(Char(car))
            car = dis.read()
        }

        return result.toString()
    }

    companion object {
        fun getPersistence(addr : String,port : Int) : Persistence {
            if(port <= 0 || port > 65535) throw IllegalArgumentException("Error: invalid port value")
            val address = InetAddress.getByName(addr)
            return Persistence(address,port)
        }
    }

    enum class Timeout(val value : Int) {
        INFINITY(0),
        STANDARD(5000)
    }
}